include/hermite_multidimensional.hpp

```diff
@@ -48,106 +48,108 @@ ullint vec2index(std::vector<int> &pos, int resolution) {
 
 }
 
+namespace libwalrus {
+
 /**
- * Returns the indices of the tensor corresponding to a given element
+ * Returns photon number statistics of a Gaussian state for a given covariance matrix `mat`.
+ * as described in *Multidimensional Hermite polynomials and photon distribution for polymode mixed light*
+ * [arxiv:9308033](https://arxiv.org/abs/hep-th/9308033).
+ *
+ * This implementation is based on the MATLAB code available at
+ * https://github.com/clementsw/gaussian-optics
  *
- * @param val
- * @param base
- * @param n
+ * @param mat a flattened vector of size \f$2n^2\f$, representing an
+ *       \f$2n\times 2n\f$ row-ordered symmetric matrix.
+ * @param d a flattened vector of size \f$2n\f$, representing the first order moments.
+ * @param resolution highest number of photons to be resolved.
  *
- * @return tensor index
  */
-std::vector<int> find_rep(int val, int base, int n) {
-    std::vector<int> x(n, 0);
-    int local_val = val;
+template <typename T>
+inline std::vector<T> hermite_multidimensional_cpp(std::vector<T> &R_mat, std::vector<T> &y_mat, int &resolution) {
+    int dim = std::sqrt(static_cast<double>(R_mat.size()));
 
-    x[0] = 1;
+    namespace eg = Eigen;
 
-    for (int i = 1; i < n; i++)
-        x[i] = x[i-1]*base;
+    eg::Matrix<T, eg::Dynamic, eg::Dynamic> R = eg::Map<eg::Matrix<T, eg::Dynamic, eg::Dynamic>, eg::Unaligned>(R_mat.data(), dim, dim);
+    eg::Matrix<T, eg::Dynamic, eg::Dynamic> y = eg::Map<eg::Matrix<T, eg::Dynamic, eg::Dynamic>, eg::Unaligned>(y_mat.data(), dim, dim);
```

Comment:
Just out of curiousity, do these two variables `R` and `y` need to be eigen matrices? I scanned through the code below, and couldn't see a spot that uses eigen methods (unless I missed it)

Reply:
You are correct. Eigen is not used here, at all.

Reply:
Might get even faster then by avoiding this conversion 😆

Reply:
This might be worth checking...